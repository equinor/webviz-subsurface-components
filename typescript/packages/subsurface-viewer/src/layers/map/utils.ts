// Rotate x,y around x0, y0 rad radians
export function rotate(
    x: number,
    y: number,
    x0: number,
    y0: number,
    rad: number
): [number, number] {
    const xRot = Math.cos(rad) * (x - x0) - Math.sin(rad) * (y - y0) + x0; // eslint-disable-line
    const yRot = Math.sin(rad) * (x - x0) + Math.cos(rad) * (y - y0) + y0; // eslint-disable-line
    return [xRot, yRot];
}
