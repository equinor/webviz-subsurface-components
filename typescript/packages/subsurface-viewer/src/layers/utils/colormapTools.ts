import type { Color } from "@deck.gl/core";

import type { SamplerProps, Texture, TextureProps } from "@luma.gl/core";

import type {
    colorTablesArray,
    createColorMapFunction as createColormapFunction,
} from "@emerson-eps/color-tables/";
import { rgbValues } from "@emerson-eps/color-tables/";
import { createDefaultContinuousColorScale } from "@emerson-eps/color-tables/dist/component/Utils/legendCommonFunction";

import type { DeckGLLayerContext } from "./layerTools";

/** Type of functions returning a color from a value in the [0,1] range. */
export type ColormapFunctionType = ReturnType<typeof createColormapFunction>;
/** @deprecated Use ColormapFunctionType instead. */
export type colorMapFunctionType = ColormapFunctionType;
/** @deprecated Use ColormapFunctionType instead. */
export type ColorMapFunctionType = ColormapFunctionType;

export interface IColormapHints {
    discreteData: boolean;
    colormapSize: number;
}

/**
 * Colormap definition, taken from the application color tables,
 * used for mapping data values to colors.
 *
 * @property colormapName - The name of the colormap.
 */
export interface ColorTableProps {
    colormapName: string;
}

/**
 * Colormap definition used for mapping data values to colors.
 *
 * @property colorTables - An array or collection of color tables from which to select the colormap.
 */
export interface ColorTableDef extends ColorTableProps {
    colorTables: colorTablesArray;
}

export type ColormapProps = ColorMapFunctionType | Uint8Array | ColorTableProps;

/**
 * Represents the possible types that can be used as colormap properties.
 *
 * - `ColormapFunctionType`: A function converting a value to a color.
 * - `Uint8Array`: A typed array containing color data.
 * - `ColorTableDef`: An object representing a color table.
 */
export type TColormapDef = ColormapFunctionType | Uint8Array | ColorTableDef;

/**
 * Creates an array of colors as RGB triplets in range [0, 1] using the colormap definition.
 * @param colormapDef Definition of the colormap.
 * @param colormapSize Number of colors in the color map.
 * @param discreteColormapFunction If true, the color map function is targeting indices ranging from 0 to colormapSize.
 * @returns Array of colors.
 */
export function getImageData(
    colormapDef: TColormapDef,
    colormapSize: number = 256,
    discreteColormapFunction: boolean = false
): Uint8Array {
    type funcType = (x: number) => Color;

    if (colormapDef instanceof Uint8Array) {
        // If colorMapProps is a Uint8Array, return it directly
        return colormapDef;
    }

    const isFunctionDefined = colormapDef !== undefined;
    const isColorTableDef =
        "colormapName" in colormapDef &&
        "colorTables" in colormapDef &&
        !!colormapDef.colorTables;

    let colormap = createDefaultContinuousColorScale() as unknown as funcType;

    if (isColorTableDef) {
        if (colormapDef.colormapName) {
            colormap = (value: number) =>
                rgbValues(
                    value,
                    colormapDef.colormapName,
                    colormapDef.colorTables
                );
        }
    } else if (isFunctionDefined) {
        colormap =
            typeof colormapDef === "function"
                ? (colormapDef as funcType)
                : ((() => colormapDef) as unknown as funcType);
    }

    const data = new Uint8Array(colormapSize * 3);

    const scaling = discreteColormapFunction
        ? 1
        : 1 / Math.max(colormapSize - 1, 1);
    for (let i = 0; i < colormapSize; i++) {
        const color = colormap ? colormap(scaling * i) : [1, 0, 0];
        if (color) {
            data[3 * i + 0] = color[0];
            data[3 * i + 1] = color[1];
            data[3 * i + 2] = color[2];
        }
    }

    return data ? data : new Uint8Array([0, 0, 0]);
}

const DEFAULT_TEXTURE_PARAMETERS: SamplerProps = {
    addressModeU: "clamp-to-edge",
    addressModeV: "clamp-to-edge",
    minFilter: "linear",
    magFilter: "linear",
};

const DISCRETE_TEXTURE_PARAMETERS: SamplerProps = {
    minFilter: "nearest",
    magFilter: "nearest",
    addressModeU: "clamp-to-edge",
    addressModeV: "clamp-to-edge",
};
export function createColormapTexture(
    colormap: TColormapDef,
    context: DeckGLLayerContext,
    colormapHints: IColormapHints
): Texture {
    const textureProps: TextureProps = {
        sampler: DEFAULT_TEXTURE_PARAMETERS,
        width: 256,
        height: 1,
        format: "rgb8unorm-webgl",
    };

    if (colormapHints.discreteData) {
        if (colormapHints.colormapSize === 0) {
            const colormapTexture = context.device.createTexture({
                ...textureProps,
                sampler: DISCRETE_TEXTURE_PARAMETERS,
                width: colormapHints.colormapSize,
                data: new Uint8Array([0, 0, 0, 0, 0, 0]),
            });
            return colormapTexture;
        }

        const colormapData =
            colormap instanceof Uint8Array
                ? colormap
                : getImageData(
                      colormap,
                      colormapHints.colormapSize,
                      colormapHints.discreteData
                  );

        const colormapTexture = context.device.createTexture({
            ...textureProps,
            sampler: DISCRETE_TEXTURE_PARAMETERS,
            width: colormapHints.colormapSize,
            data: colormapData,
        });

        return colormapTexture;
    }

    const data = getImageData(colormap);

    const colormapTexture = context.device.createTexture({
        ...textureProps,
        data: data,
    });
    return colormapTexture;
}
