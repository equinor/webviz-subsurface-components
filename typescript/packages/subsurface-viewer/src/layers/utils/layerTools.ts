import { PickingInfo } from "@deck.gl/core/typed";
import { Color } from "@deck.gl/core/typed";
import {
    Layer,
    LayersList,
    LayerManager,
    CompositeLayerProps,
} from "@deck.gl/core/typed";
import { Matrix4 } from "math.gl";
import {
    ContinuousLegendDataType,
    DiscreteLegendDataType,
} from "../../components/ColorLegend";
import DrawingLayer from "../drawing/drawingLayer";

export type Position3D = [number, number, number];

// Return a color given a number in the [0,1] range.
export type colorMapFunctionType = (x: number) => [number, number, number];

export interface ExtendedLayerProps<D> extends CompositeLayerProps<D> {
    "@@type"?: string;
    name: string;
}

export interface ExtendedLayer<D> extends Layer<D> {
    getLegendData?: () => DiscreteLegendDataType | ContinuousLegendDataType;
}

export interface PropertyDataType {
    name: string;
    value: string | number;
    color?: Color;
}

// Layer pick info can have multiple properties
export interface LayerPickInfo extends PickingInfo {
    propertyValue?: number; // for single property
    properties?: PropertyDataType[]; // for multiple properties
}

// Creates property object which will be used to display layer property
// in the info card.
export function createPropertyData(
    name: string,
    value: string | number,
    color?: Color
): PropertyDataType {
    return {
        name: name,
        value: value,
        color: color,
    };
}

// Return a model matrix representing a rotation of "deg" degrees around the point x, y
export function getModelMatrix(deg: number, x: number, y: number): Matrix4 {
    const rad = deg * 0.017453;
    const IDENTITY = [1, 0, 0, 0, 0, 1, 0, 0, 0, 0, 1, 0, 0, 0, 0, 1];

    const m1 = new Matrix4(IDENTITY).translate([-x, -y, 0, 1]); // translate to origin
    const mRot = new Matrix4(IDENTITY).rotateZ(rad); // rotate
    const m2 = new Matrix4(IDENTITY).translate([x, y, 0, 1]); // translate back

    // Make  m2*mRot*m1
    mRot.multiplyRight(m1);
    const m2mRotm1 = m2.multiplyRight(mRot);

    return m2mRotm1;
}

// Return a model matrix representing a rotation of "deg" degrees around the point x, y
export function getModelMatrixScale(scaleZ: number): Matrix4 {
    const IDENTITY = [1, 0, 0, 0, 0, 1, 0, 0, 0, 0, 1, 0, 0, 0, 0, 1];
    const mScaleZ = new Matrix4(IDENTITY).scale([1, 1, scaleZ]);
    return mScaleZ;
}

export function getLayersInViewport(
    layers: Record<string, unknown>[] | LayersList,
    layerIds: string[] | undefined
): Record<string, unknown>[] | LayersList {
    if (layerIds && layerIds.length > 0 && layers) {
        const layers_in_view = (layers as never[]).filter((layer) =>
            layerIds.includes(layer["id"] as string)
        );
        return layers_in_view;
    } else {
        return layers;
    }
}

export function getLayersByType(layers: LayersList, type: string): LayersList {
    if (!layers) return [];
    return layers.filter((l) => l?.constructor.name === type);
}

export type NewLayersList = LayersList & {
    id: string;
    props: prop;
};

type prop = {
    data: wellData;
    visible: boolean;
};

type wellData = {
    features: feature[];
    type: string;
    unit?: string;
};

type feature = {
    properties: {
        name: string;
    };
};

export function getWellLayerByTypeAndSelectedWells(
    layers: LayersList,
    type: string,
    selectedWell: string
): LayersList {
    if (!layers) return [];
    return layers.filter((l) => {
        return (
            l?.constructor.name === type &&
            (l as NewLayersList).props.data.features.find(
                (item) => item.properties.name === selectedWell
            )
        );
    });
}

export function getLayersById(layers: LayersList, id: string): LayersList {
    if (!layers) return [];
    return layers.filter((l) => (l as Layer).id === id);
}

export function isDrawingEnabled(layer_manager: LayerManager): boolean {
    const drawing_layer = layer_manager.getLayers({
        layerIds: ["drawing-layer"],
    })?.[0] as DrawingLayer;
    return (
        drawing_layer &&
        drawing_layer.props.visible &&
        drawing_layer.props.mode != "view"
    );
}

export function invertZCoordinate(dataArray: Float32Array): void {
    for (let i = 2; i < dataArray.length; i += 3) {
        dataArray[i] *= -1;
    }
}

export function defineBoundingBox(
    dataArray: Float32Array
): [number, number, number, number, number, number] {
    const length = dataArray.length;
    let minX = Number.POSITIVE_INFINITY;
    let minY = Number.POSITIVE_INFINITY;
    let minZ = Number.POSITIVE_INFINITY;
    let maxX = Number.NEGATIVE_INFINITY;
    let maxY = Number.NEGATIVE_INFINITY;
    let maxZ = Number.NEGATIVE_INFINITY;

    for (let i = 0; i < length; i += 3) {
        const x = dataArray[i];
        const y = dataArray[i + 1];
        const z = dataArray[i + 2];
        minX = x < minX ? x : minX;
        minY = y < minY ? y : minY;
        minZ = z < minZ ? z : minZ;

        maxX = x > maxX ? x : maxX;
        maxY = y > maxY ? y : maxY;
        maxZ = z > maxZ ? z : maxZ;
    }
    return [minX, minY, minZ, maxX, maxY, maxZ];
}
