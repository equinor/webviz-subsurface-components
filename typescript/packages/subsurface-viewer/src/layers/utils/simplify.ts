// (c) 2013, Vladimir Agafonkin
// https://github.com/mourner/simplify-js/blob/3d/simplify.js
// Modified version of above libary. Port to Typescript. Use subset. Take into account md values.

import type { Position3D } from "../utils/layerTools";

// square distance from a point to a segment
function getSquareSegmentDistance(
    p: Position3D,
    p1: Position3D,
    p2: Position3D
) {
    let x = p1[0],
        y = p1[1],
        z = p1[2],
        dx = p2[0] - x,
        dy = p2[1] - y,
        dz = p2[2] - z;

    if (dx !== 0 || dy !== 0 || dz !== 0) {
        const t =
            ((p[0] - x) * dx + (p[1] - y) * dy + (p[2] - z) * dz) /
            (dx * dx + dy * dy + dz * dz);

        if (t > 1) {
            x = p2[0];
            y = p2[1];
            z = p2[2];
        } else if (t > 0) {
            x += dx * t;
            y += dy * t;
            z += dz * t;
        }
    }

    dx = p[0] - x;
    dy = p[1] - y;
    dz = p[2] - z;

    return dx * dx + dy * dy + dz * dz;
}

// simplification using optimized Douglas-Peucker algorithm with recursion elimination
function simplifyDouglasPeucker(
    points: Position3D[],
    mds: number[],
    sqTolerance: number
) {
    if (points.length <= 2 || 0 === sqTolerance) {
        return [points, mds];
    }

    const len = points.length,
        MarkerArray = typeof Uint8Array !== "undefined" ? Uint8Array : Array,
        markers = new MarkerArray(len),
        stack = [],
        newPoints = [],
        newMds = [];
    let first = 0,
        last = len - 1,
        i = 0,
        maxSqDist,
        sqDist,
        index = 0;
    markers[first] = markers[last] = 1;

    while (last) {
        maxSqDist = 0;

        for (i = first + 1; i < last; i++) {
            sqDist = getSquareSegmentDistance(
                points[i],
                points[first],
                points[last]
            );

            if (sqDist > maxSqDist) {
                index = i;
                maxSqDist = sqDist;
            }
        }

        if (maxSqDist > sqTolerance) {
            markers[index] = 1;
            stack.push(first, index, index, last);
        }

        last = stack.pop() as number;
        first = stack.pop() as number;
    }

    for (i = 0; i < len; i++) {
        if (markers[i]) {
            newPoints.push(points[i]);
            newMds.push(mds[i]);
        }
    }

    return [newPoints, newMds];
}

/**
 * Simplify points based on the given tolerance
 * @param points Points to simplify
 * @param mds MD values
 * @param tolerance The tolerance value to determine the level of simplification. A higher tolerance results in fewer points. If undefined or zero, then no simplification is applied.
 * @returns Simplified points and mds
 */
export function simplify(
    points: Position3D[],
    mds: number[],
    tolerance?: number
) {
    const sqTolerance = tolerance !== undefined ? tolerance * tolerance : 0;
    const [newPoints, newMds] = simplifyDouglasPeucker(
        points,
        mds,
        sqTolerance
    );
    return [newPoints, newMds];
}
