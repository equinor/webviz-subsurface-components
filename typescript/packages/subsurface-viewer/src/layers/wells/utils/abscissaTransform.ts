import type { FeatureCollection, GeometryCollection, Position } from "geojson";
import { cloneDeep, zip } from "lodash";
import { distance } from "mathjs";

function computeUnfoldedPath(worldCoordinates: Position[]): Position[] {
    const z = worldCoordinates.map((v) => v[2]);
    const delta = worldCoordinates.map((v, i, coordinates) => {
        const prev = coordinates[i - 1] || v;
        return distance([prev[0], prev[1]], [v[0], v[1]]);
    });
    const a: number[] = [];
    delta.forEach((d) => {
        const prev = a.at(-1) || 0;
        a.push((d as number) + prev);
    });
    const vAbscissa = zip(a, z, [...a].fill(0));

    return vAbscissa as Position[];
}

/** Projects well trajectories unfolded onto an XY plane.  */
export function abscissaTransform(
    featureCollection: FeatureCollection<GeometryCollection>
) {
    const featureCollectionCopy = cloneDeep(featureCollection);
    for (const feature of featureCollectionCopy.features) {
        const geometryCollection = feature.geometry;
        for (const geometry of geometryCollection.geometries) {
            if ("LineString" === geometry.type) {
                const transformedCoordinates = computeUnfoldedPath(
                    geometry.coordinates
                );
                geometry.coordinates = transformedCoordinates;
            } else if ("Point" === geometry.type) {
                const coordinates = geometry.coordinates;
                geometry.coordinates = [0, coordinates[2], 0];
            }
        }
    }

    return featureCollectionCopy;
}
