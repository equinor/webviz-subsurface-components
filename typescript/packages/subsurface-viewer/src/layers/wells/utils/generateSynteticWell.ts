export function generateSynteticWell() {
    const well = {
        type: "FeatureCollection",
        features: [
            {
                type: "Feature",
                geometry: {
                    type: "GeometryCollection",
                    geometries: [
                        {
                            type: "Point",
                            coordinates: [0, 0],
                        },
                        {
                            type: "LineString",
                            coordinates: [
                                [0, 0, 0],
                                [0, 0, -100],
                                [0, 0, -200],
                                [0, 0, -200],
                                [0, 0, -200],
                                [0, 0, -300],
                                [0, 0, -400],
                                [0, 0, -400],
                                [0, 0, -500],
                            ],
                        },
                    ],
                },
                properties: {
                    name: "wl6",
                    color: [255, 255, 0, 255],
                    md: [[0, 1, 2, 3, 4, 5, 8, 9]],
                },
            },
        ],
    };

    return well;
}
