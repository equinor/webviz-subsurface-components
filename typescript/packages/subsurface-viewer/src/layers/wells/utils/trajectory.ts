import type { Color, Position } from "@deck.gl/core";
import type { Feature, GeometryCollection, LineString } from "geojson";
import type { StyleAccessorFunction } from "../../types";

type ColorAccessor = Color | StyleAccessorFunction | undefined;

function getLineStringGeometry(well_object: Feature): LineString {
    return (well_object.geometry as GeometryCollection)?.geometries.find(
        (item: { type: string }) => item.type == "LineString"
    ) as LineString;
}

export function getColor(accessor: ColorAccessor) {
    if (accessor as Color) {
        return accessor as Color;
    }

    return (object: Feature, objectInfo?: Record<string, unknown>): Color => {
        if (typeof accessor === "function") {
            const color = (accessor as StyleAccessorFunction)(
                object,
                objectInfo
            ) as Color;
            if (color) {
                return color;
            }
        }
        return object.properties?.["color"] as Color;
    };
}

/**
 * Get trajectory transparency based on alpha of trajectory color
 */
function isTrajectoryTransparent(
    well_object: Feature,
    color_accessor: ColorAccessor
): boolean {
    let alpha;
    const accessor = getColor(color_accessor);
    if (typeof accessor === "function") {
        alpha = accessor(well_object)?.[3];
    } else {
        alpha = (accessor as Color)?.[3];
    }
    return alpha === 0;
}

/**
 * Get trajectory data from LineString Geometry if it's visible (checking
 * trajectory visiblity based on line color)
 */
export function getTrajectory(
    well_object: Feature,
    color_accessor: ColorAccessor
): Position[] | undefined {
    if (!isTrajectoryTransparent(well_object, color_accessor))
        return getLineStringGeometry(well_object)?.coordinates as Position[];
    else return undefined;
}
