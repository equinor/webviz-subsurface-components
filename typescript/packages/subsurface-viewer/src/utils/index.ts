export type { BoundingBox2D, Point2D } from "./BoundingBox2D";
export type { BoundingBox3D } from "./BoundingBox3D";
export type { Point3D } from "./Point3D";
export { add as addPoints3D } from "./Point3D";

export { proportionalZoom, scaleZoom } from "./camera";

export { useScaleFactor } from "./event";
