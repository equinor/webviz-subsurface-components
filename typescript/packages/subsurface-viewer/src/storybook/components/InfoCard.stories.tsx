import type { Meta, StoryObj } from "@storybook/react";
import React from "react";

import InfoCard from "../../components/InfoCard";
import { type LayerPickInfo } from "../../layers/utils/layerTools";

const stories: Meta = {
    component: InfoCard,
    title: "SubsurfaceViewer/Components/InfoCard",
};
export default stories;

export const SingleProperty: StoryObj<typeof InfoCard> = {
    args: {
        pickInfos: [
            {
                x: 152,
                y: 254,
                coordinate: [111, 222],
            } as LayerPickInfo,
        ],
    },
    render: (args) => <InfoCard {...args} />,
};

export const MutipleProperties: StoryObj<typeof InfoCard> = {
    args: {
        pickInfos: [
            {
                x: 152,
                y: 254,
                coordinate: [111, 222],
            } as LayerPickInfo,
            {
                layer: { id: "wells-layer" },
                property: { name: "Poro WellA", value: 123 },
            },
        ],
    },
};
