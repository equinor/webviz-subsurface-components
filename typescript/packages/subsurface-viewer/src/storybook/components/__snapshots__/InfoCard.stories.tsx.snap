// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SubsurfaceViewer/Components/InfoCard MutipleProperties smoke-test 1`] = `
<div class="MuiTableContainer-root css-kge0eu">
  <table class="MuiTable-root css-2xc83k"
         aria-label="info-card"
  >
    <tbody class="MuiTableBody-root css-1xnox0e">
      <tr class="MuiTableRow-root css-1tfy527">
        <td class="MuiTableCell-root MuiTableCell-body MuiTableCell-sizeMedium css-1nx2l6">
          <button type="button"
                  tabindex="0"
                  class="Button__ButtonBase-sc-1hs0myn-1 eVWetT"
                  style="padding: 0px;"
          >
            <span class="Button__Inner-sc-1hs0myn-0 tfpDC">
              <svg fill="currentColor"
                   viewbox="0 0 24 24"
                   aria-hidden="true"
                   xmlns="http://www.w3.org/2000/svg"
                   height="24px"
                   width="24px"
                   class="Icon__StyledSvg-sc-6evbi1-0 jjFape"
              >
                <path data-testid="eds-icon-path"
                      d="m7 14.5 5-5 5 5H7Z"
                      fill-rule="evenodd"
                      clip-rule="evenodd"
                      class="Icon__StyledPath-sc-6evbi1-1 erHlfr"
                >
                </path>
              </svg>
            </span>
          </button>
        </td>
        <td class="MuiTableCell-root MuiTableCell-body MuiTableCell-sizeMedium css-q34dxg">
          Position
        </td>
      </tr>
      <tr class="MuiTableRow-root css-1tfy527">
        <td class="MuiTableCell-root MuiTableCell-body MuiTableCell-sizeMedium css-q34dxg"
            colspan="2"
            style="padding-bottom: 0px; padding-top: 0px;"
        >
          <div class="MuiCollapse-root MuiCollapse-vertical MuiCollapse-entered css-c4sutr"
               style="min-height: 0px;"
          >
            <div class="MuiCollapse-wrapper MuiCollapse-vertical css-hboir5">
              <div class="MuiCollapse-wrapperInner MuiCollapse-vertical css-8atqhb">
                <table class="MuiTable-root css-1owb465"
                       aria-label="properties"
                >
                  <tbody class="MuiTableBody-root css-1xnox0e">
                    <tr class="MuiTableRow-root css-1tfy527">
                      <td class="MuiTableCell-root MuiTableCell-body MuiTableCell-sizeSmall css-1o6fzn1"
                          style="border: none; padding-left: 10px; padding-right: 10px;"
                      >
                        x
                      </td>
                      <td class="MuiTableCell-root MuiTableCell-body MuiTableCell-sizeSmall css-1o6fzn1"
                          style="border: none; text-align: right;"
                      >
                        111.00 m
                      </td>
                    </tr>
                    <tr class="MuiTableRow-root css-1tfy527">
                      <td class="MuiTableCell-root MuiTableCell-body MuiTableCell-sizeSmall css-1o6fzn1"
                          style="border: none; padding-left: 10px; padding-right: 10px;"
                      >
                        y
                      </td>
                      <td class="MuiTableCell-root MuiTableCell-body MuiTableCell-sizeSmall css-1o6fzn1"
                          style="border: none; text-align: right;"
                      >
                        222.00 m
                      </td>
                    </tr>
                  </tbody>
                </table>
              </div>
            </div>
          </div>
        </td>
      </tr>
    </tbody>
  </table>
</div>
`;

exports[`SubsurfaceViewer/Components/InfoCard SingleProperty smoke-test 1`] = `
<div class="MuiTableContainer-root css-kge0eu">
  <table class="MuiTable-root css-2xc83k"
         aria-label="info-card"
  >
    <tbody class="MuiTableBody-root css-1xnox0e">
      <tr class="MuiTableRow-root css-1tfy527">
        <td class="MuiTableCell-root MuiTableCell-body MuiTableCell-sizeMedium css-1nx2l6">
          <button type="button"
                  tabindex="0"
                  class="Button__ButtonBase-sc-1hs0myn-1 eVWetT"
                  style="padding: 0px;"
          >
            <span class="Button__Inner-sc-1hs0myn-0 tfpDC">
              <svg fill="currentColor"
                   viewbox="0 0 24 24"
                   aria-hidden="true"
                   xmlns="http://www.w3.org/2000/svg"
                   height="24px"
                   width="24px"
                   class="Icon__StyledSvg-sc-6evbi1-0 jjFape"
              >
                <path data-testid="eds-icon-path"
                      d="m7 14.5 5-5 5 5H7Z"
                      fill-rule="evenodd"
                      clip-rule="evenodd"
                      class="Icon__StyledPath-sc-6evbi1-1 erHlfr"
                >
                </path>
              </svg>
            </span>
          </button>
        </td>
        <td class="MuiTableCell-root MuiTableCell-body MuiTableCell-sizeMedium css-q34dxg">
          Position
        </td>
      </tr>
      <tr class="MuiTableRow-root css-1tfy527">
        <td class="MuiTableCell-root MuiTableCell-body MuiTableCell-sizeMedium css-q34dxg"
            colspan="2"
            style="padding-bottom: 0px; padding-top: 0px;"
        >
          <div class="MuiCollapse-root MuiCollapse-vertical MuiCollapse-entered css-c4sutr"
               style="min-height: 0px;"
          >
            <div class="MuiCollapse-wrapper MuiCollapse-vertical css-hboir5">
              <div class="MuiCollapse-wrapperInner MuiCollapse-vertical css-8atqhb">
                <table class="MuiTable-root css-1owb465"
                       aria-label="properties"
                >
                  <tbody class="MuiTableBody-root css-1xnox0e">
                    <tr class="MuiTableRow-root css-1tfy527">
                      <td class="MuiTableCell-root MuiTableCell-body MuiTableCell-sizeSmall css-1o6fzn1"
                          style="border: none; padding-left: 10px; padding-right: 10px;"
                      >
                        x
                      </td>
                      <td class="MuiTableCell-root MuiTableCell-body MuiTableCell-sizeSmall css-1o6fzn1"
                          style="border: none; text-align: right;"
                      >
                        111.00 m
                      </td>
                    </tr>
                    <tr class="MuiTableRow-root css-1tfy527">
                      <td class="MuiTableCell-root MuiTableCell-body MuiTableCell-sizeSmall css-1o6fzn1"
                          style="border: none; padding-left: 10px; padding-right: 10px;"
                      >
                        y
                      </td>
                      <td class="MuiTableCell-root MuiTableCell-body MuiTableCell-sizeSmall css-1o6fzn1"
                          style="border: none; text-align: right;"
                      >
                        222.00 m
                      </td>
                    </tr>
                  </tbody>
                </table>
              </div>
            </div>
          </div>
        </td>
      </tr>
    </tbody>
  </table>
</div>
`;
